worker/src/monitor/metrics_collector.rs

```diff
@@ -0,0 +1,67 @@
+use std::{thread::sleep, time::Duration};
+
+use proto::common::node::Metrics;
+use sysinfo::System;
+
+const CPU_DELAY_IN_MILLIS: u64 = 500;
+
+/// Responsable for collecting metrics about the System, such as
+/// CPU Percentage Usage, Free Memory Space.
+pub struct MetricsCollector {
+    system: System,
+}
+impl MetricsCollector {
+    /// Instantiates a new `MetricsCollector`.
+    pub fn new() -> Self {
+        Self {
+            system: System::new_all(),
+        }
+    }
+
+    /// Get the `cpu_usages` and `free_memory` metrics from the current system.
+    ///
+    /// When calling this method, will sleep the thread for
+    /// `CPU_DELAY_IN_MILLIS`.
+    pub fn get_metrics(&mut self) -> Metrics {
+        Metrics {
+            cpu_usage: self.get_cpu_usage(),
+            mem_total_mib: self.get_total_memory(),
+            mem_used_mib: self.get_used_memory(),
+        }
+    }
+
+    fn get_cpu_usage(&mut self) -> f64 {
+        self.system.refresh_cpu();
+        sleep(Duration::from_millis(CPU_DELAY_IN_MILLIS));
```

Comment:
As we're using async, we should not use a sync sleep here.

We have two options:
1. Make this function (and all of its callers) `async` so that we may call tokio's `sleep` here. In this case, we'd also need to parameterize the delay since it's going to be dynamic from the command-line arguments.
2. Remove the sleep from this function by delegating it to the caller. In this case, the caller is `main`, which _should_ only call `get_metrics` after specific intervals of time.

Personally, I'm more fond of the latter option.

Reply:
I also vouch for option 2 here

Reply:
But how to make main call `get_metrics` within these intervals without a `loop` + `sleep` combination? How would that be structured?

Reply:
The loop and a sleep (from tokio) would still exist. The point is that they wouldn’t be on the colletor module, but rather on one of its calling modules.